CONTRIBUTING.md

```diff
@@ -2,32 +2,97 @@
 
-_Note: You do not have to read the following information to contribute._
+Next you will push your commit to GitHub, which means add it to `username/sandbox` on GitHub. To do this, run:
 
-## Commit Access
+```sh
+# This will only have to be run once per computer
+# It adopts the Git 2.0, removing a confusing message
+git config --global push.default simple
+
+# Push your current branch to GitHub
+# If this is the first time you're pushing this branch,
+# you will receive an error suggesting you specify --set-upstream.
+# If so, use the suggested command. Then future git push attempts will work.
+git push
+```
+
+Now you are ready to create your [pull request](https://help.github.com/articles/about-pull-requests/). The GitHub webpage for either `cognoma/sandbox` or `username/sandbox` should have a new button to "Compare & pull request." Click it and write a title and description for your requested changes. Then submit. Soon (hopefully), another Cognoma contributor will review your pull request. Usually, pull requests will require some changes before they are ready to merge (incorporate into Cognoma repository). You can add additional changes to a pull request by adding commits to the branch its based on, which is this example is `add-username`.
+
+Once your pull request passes review, it will be merged by a repository maintainer. Before starting on any new pull requests, you will want to sync your fork, so you're not starting from an outdated version of the codebase, by running:
+
+```
+# Fetch all new changes on the cognoma repo
+git fetch upstream
+
+# Make sure you're on your master branch.
+git checkout master
+
+# Fast forward your master branch with upstream changes.
+git merge --no-ff upstream/master
+```
+
+### Pull request advice
+
+1. Focus a pull request on a single feature. Smaller pull requests are easier to review and will be thus get merged quicker. If you want to change multiple things, create multiple pull requests.
++ Open incomplete pull requests but put WIP in the title for "work in progress". This enables early stage feedback to save time later and helps coordinate efforts. Once you feel the pull request is ready, remove WIP from the title.
++ Please help review open pull requests. Even if you don't fully know the topic, the dialogue will often be productive. As a reviewer, it's helpful to note the type of review you performed: did you test the code, look it over, or are you just supporting the concept?
++ Repository maintainers are encouraged to squash merge pull requests when appropriate. This will compact all of the commits into one, heling keep the commit history clean.
```

Comment:
a [link](https://github.com/blog/2141-squash-your-commits) here would probably be good